SLA-to-Deadlines migration guide and minor Deadlines-related docs updates (#55743)
mistercrunch/Airflow: airflow-core/docs/howto/sla-to-deadlines.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Migrating from SLA to Deadline Alerts
=====================================

> Two different paradigms

Reviewer: I think you'd gone with caps titles in the other docs.
```suggestion
Two Different Paradigms
```

-----------------------

> While the goal of the **SLA** and **Deadlines** features are very similar, they use two very different approaches

Reviewer: Match the name in the title for the first usage?

```suggestion
While the goal of the **SLA** and **Deadline Alerts** features are very similar, they use two very different approaches
```

Author: Made the change here and elsewhere in the file, thanks

> which makes it difficult to present a universal migration guide. Instead, this guide will lay out the major

Reviewer: Don't under sell yourself in the docs. I know you want a hedge, but I think it sounds more official and trustworthy if we don't IMHO.

```suggestion
While the goal of the **SLA** and **Deadline Alerts** features are very similar, they use two very different approaches, this guide will lay out the major
```

differences and help you decide on the best approach for your use case.

To begin with, we'll start by explaining the two approaches then go into how to find the right Deadline for your use case.

SLA
^^^

When the dag run **finishes**, check the current time.  If the time is greater than (logical_date + sla) then
execute ``sla_miss_callback``.  If the Dag run never finishes, the SLA is never checked.

Deadlines
^^^^^^^^^

When a Dag run **starts**, calculate and store (:ref:`DeadlineReference <built-in-deadline-references>` + interval).
Every pass of the scheduler loop (default 5 seconds, set by ``scheduler_heartbeat_sec``) if any of those times have

> passed then execute ``callback(**kwargs)``.

Reviewer: Suggestion, feel free to ignore.

```suggestion
The scheduler loop then periodically checks (default 5 seconds, set by ``scheduler_heartbeat_sec``) if any of those times have
passed then execute ``callback(**kwargs)``.
```

The most direct migration path would be to use the ``DeadlineReference.DAGRUN_LOGICAL_DATE`` reference, but note that
the major change is that the Deadline's callback will execute "immediately" (within ``scheduler_heartbeat_sec`` of the

> calculated expiation time) and not wait until the Dag finishes first.

Reviewer 2: expiration

Author: I wonder why spell_check didn't catch that.   Cheers!

Equivalent Example Dags
-----------------------

Below is a Dag using a 1-hour SLA, followed by an equivalent Dag using Deadline Alerts.

SLA Example
^^^^^^^^^^^

.. code-block:: python

  with DAG(
      "minimal_sla_example",
      default_args={"sla": timedelta(hours=1)},
      sla_miss_callback=SlackWebhookNotifier(
          text="SLA missed for {{ dag.dag_id }}",
      ),
  ):
      BashOperator(task_id="long_task", bash_command="sleep 3600")

Deadlines Example
^^^^^^^^^^^^^^^^^

.. code-block:: python

  with DAG(
      "minimal_deadline_example",
      deadline=DeadlineAlert(
          reference=DeadlineReference.DAGRUN_LOGICAL_DATE,
          interval=timedelta(hours=1),
          callback=AsyncCallback(
              SlackWebhookNotifier,
              kwargs={

>                   text: "Deadline missed for {{ dag.dag_id }}",

Reviewer 3: Since we don't have Dag in the minimal context currently supplied to the callbacks, this would need to use the dag run to get the dag_id.
```suggestion
                  text: "Deadline missed for {{ dag_run.dag_id }}",
```
Maybe we should use dag_run for both examples (it should work in the SLA example as well) and include a   disclaimer that links here?
https://github.com/apache/airflow/blob/7b35e457be19b416589e20bfa9653cd59c29d19b/airflow-core/docs/howto/deadline-alerts.rst?plain=1#L218-L226

Author: Of course, good catch, thanks!

              },
          ),
      ),
  ):
      BashOperator(task_id="long_task", bash_command="sleep 3600")

> ## Further Reading

Reviewer 2: This didn't render correctly as a heading

Author: That's because I did it wrong :P    Thanks
